4.5 Data Consistency Checks

In [8]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [10]:
#create path
path =  r'C:\Users\Marker\Videos\Instakart Basket Analysis 01-25'
#orders data
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original data', 'orders.csv'), index_col = False)
#products data
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original data', 'products.csv'), index_col = False)
#departments data
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original data', 'departments.csv'), index_col = False)

Mixed-Type Data

In [7]:
# Create a dataframe
df_test = pd.DataFrame()

In [9]:
#Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [11]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [13]:
# Check for any mixed-type columns
for col in df_test.columns.tolist():
    weird = (df_test[col].map(type) != df_test[col].iloc[0].__class__).any()
    if weird:
        print(col)

mix


In [15]:
# Convert column's data from numeric to string
df_test['mix'] = df_test['mix'].astype('str')

In [17]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   mix     4 non-null      object
dtypes: object(1)
memory usage: 164.0+ bytes


In [19]:
#Another method
df_test['mix'].dtype

dtype('O')

 Missing Values

In [21]:
#Find missing values (takes sum of Trues in a  column)
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [23]:
#Create subset of the dataframe that contains the nulls
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [25]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [27]:
#Addressing Missing Values
#There a few ways to deal with missing data:
#Create a new variable that acts like a flag based on the missing value.
#Impute the value with the mean or median of the column (if the variable is numeric).
#Remove or filter out the missing data.

In [29]:
df_nan.describe()

product_id    aisle_id  department_id     prices
count     16.000000   16.000000      16.000000  16.000000
mean    6684.000000   89.937500      10.937500  13.012500
std    12836.665242   33.731229       4.639953   3.881731
min       34.000000   26.000000       1.000000   1.200000
25%      459.250000   70.750000       7.750000  12.175000
50%     2413.000000   98.500000      11.500000  13.650000
75%     3872.750000  120.000000      14.500000  14.425000
max    40440.000000  126.000000      16.000000  20.900000

In [31]:
#When using the mean, df['column with missings'].fillna(mean value, inplace=True)

In [33]:
df_nan.median()

product_id       2413.0
product_name        NaN
aisle_id           98.5
department_id      11.5
prices            13.65
dtype: object

In [35]:
#Because the missing values are strings there's not much you can do other than remove/filter the data

In [37]:
# Find shape of df
df_prods.shape

(49693, 5)

In [39]:
# Create new df without the nulls
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [41]:
df_prods.shape

(49693, 5)

In [ ]:
#To drop all missing values --> df_prods.dropna(inplace = True)

#To drop only the NaNs from a particular column --> df_prods.dropna(subset = [‘product_name’], inplace = True)

 Duplicates

In [43]:
# Create subset of df_prods_clean that contains only rows of duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [45]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [47]:
# Check current number of rowns in df_prods_clean
df_prods_clean.shape

(49677, 5)

In [49]:
# Create datafram that doesn't include the duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [51]:
df_prods_clean_no_dups.shape

(49672, 5)

 Exporting Changes

Steps


In [55]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

Step 2

In [57]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

The column 'days_since prior order' has a count less than all other rows, indicating missing values. It also has a measure of zero as a minimum, indicating that some orders could be places within the same day. The maximum value is 30 (3.00000000e+01), which may indicate that orders are capped at 30 days or 1 month. The column 'order_number': Has a max value of 100 (1.000000e+02), which may mean orders are capped at 100 . The column 'order_dow' has a minimum value of 0 and maximim value of 6 meaning the seven numbers (0-6) represent the days of the week. The column 'order_hour_of_day' has a minimum value of 0 and maximum value of 23 meaning the twenty-four numbers (0-23) represent the 24 hrs of the day.

Step 3

In [12]:
for col in df_ords.columns.tolist():
    weird = (df_ords[col].map(type) != df_ords[col].iloc[0].__class__).any()
    if weird:
        print(col)
        

order_id
user_id
order_number
order_dow
order_hour_of_day
days_since_prior_order


In [14]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(5), object(1)
memory usage: 182.7+ MB


Step 5


In [18]:
# Find missing values (takes sum of Trues in a column)
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

As suspected, the variable that has missing values is the 'days_since_prior_order' which has 206209 values missing. In all these cases I suspect it to just one order in total, so in these cases the 'days_since_prior_order' value is missing.

Step 6

In [20]:
# Create subset of the dataframe that contains the nulls
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

This would be the way to adress the missing values but there is no need to remove these missing values, since each unique user_id should have only one missing value in the days_since_prior_order column and that missing value belongs to the very first order of each unique user_id, since they only placed one order.


In [ ]:
Step 7

In [44]:
# Create subset of df_ords that contains only rows of duplicates
df_ords_dup = df_ords[df_ords.duplicated()]

In [48]:
df_ords_dup

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []



There are no duplicates in this set because it returned empty. however, if there were duplicates I would execute 
'df_ords-dup = df_ords.drop_duplicates()' to creat a new datafreame without duplicates.

Step 9

In [50]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))